invokeai/app/invocations/ip_adapter.py

```diff
@@ -89,17 +89,32 @@ def invoke(self, context: InvocationContext) -> IPAdapterOutput:
         assert isinstance(ip_adapter_info, IPAdapterConfig)
         image_encoder_model_id = ip_adapter_info.image_encoder_model_id
         image_encoder_model_name = image_encoder_model_id.split("/")[-1].strip()
-        image_encoder_models = context.models.search_by_attrs(
-            name=image_encoder_model_name, base=BaseModelType.Any, type=ModelType.CLIPVision
-        )
-        assert len(image_encoder_models) == 1
+        image_encoder_model = self._get_image_encoder(context, image_encoder_model_name)
         return IPAdapterOutput(
             ip_adapter=IPAdapterField(
                 image=self.image,
                 ip_adapter_model=self.ip_adapter_model,
-                image_encoder_model=ModelIdentifierField.from_config(image_encoder_models[0]),
+                image_encoder_model=ModelIdentifierField.from_config(image_encoder_model),
                 weight=self.weight,
                 begin_step_percent=self.begin_step_percent,
                 end_step_percent=self.end_step_percent,
             ),
         )
+
+    def _get_image_encoder(self, context: InvocationContext, image_encoder_model_name: str) -> AnyModelConfig:
+        found = False
+        while not found:
+            image_encoder_models = context.models.search_by_attrs(
+                name=image_encoder_model_name, base=BaseModelType.Any, type=ModelType.CLIPVision
+            )
+            found = len(image_encoder_models) > 0
+            if not found:
+                context.logger.warning(
+                    f"The image encoder required by this IP Adapter ({image_encoder_model_name}) is not installed."
+                )
+                context.logger.warning("Downloading and installing now. This may take a while.")
+                installer = context._services.model_manager.install
+                job = installer.heuristic_import(f"InvokeAI/{image_encoder_model_name}")
+                installer.wait_for_job(job, timeout=600)  # wait up to 10 minutes - then raise a TimeoutException
```

Comment:
What if my connection is slow and it takes longer than 10 minutes to download it?

Reply:
Just as an example, https://huggingface.co/InvokeAI/ip-adapter-plus_sdxl_vit-h/tree/main is about 1GB. I'm at my in-laws' place today, in a rural area. It's entirely possible that this will time out for me here, especially on a bad internet day.

Reply:
Then it raises an exception and the generation fails. What is good alternative behavior? Options are: (1) hang indefinitely until user cancels the generation; or (2) lengthen the timeout.

Reply:
This is probably not the right place for a change like this, but I think a max retries makes more sense than a timeout.

Of those two options, I'd say hang indefinitely. If I have a slow connection and I want to run an IP adapter, I'm probably willing to wait as long as my internet takes to download it.

Reply:
I just tested starting the generation and then hitting the cancel button. The image generation stopped, but model installation continued to run in the background (I could see the progress bar moving in the model manager status area). This seems like pretty good behavior to me.

Reply:
Yeah, sounds fine to me - definitely good enough until we get the dependent installation logic in place.